Feature/unit tests

dneukirchen opened the pull request:
This PR brings the unit tests back to the core as discussed in https://github.com/joomla/joomla-cms/pull/23962

* added psr-4 autoloading for tests using composer 
* namespaced unit tests
* added unit tests 
* refactored existing unit tests
* removed useless unit tests
* re-integrated some of the unit tests that are currently not running
* added UnitTestCase as a base class
* updated build tools
* fixed one bug in Mail class which surprisingly was not detected by previous unit tests 

before: OK (123 tests, 144 assertions)
after: Tests: 443, Assertions: 678, Skipped: 1.

Testing instructions:
Code review + run phpunit

dneukirchen committed:
Move unit tests back to repository

dneukirchen committed:
Update config files

dneukirchen committed:
Namespace JHtmlTests

dneukirchen committed:
Delete unused html test files

dneukirchen committed:
Add base unit test case

dneukirchen committed:
Namespace Cms Tests

dneukirchen committed:
Psr-4 autoload tests

dneukirchen committed:
Refactoring JLayoutBaseTest

dneukirchen committed:
Delete JPaginationObjectTest

We do not test constructors! ;-D

dneukirchen committed:
Refactor JToolbarButtonCustomTest

dneukirchen committed:
Remove JToolbarButtonTest

We should not test constructor params and getters/setters

dneukirchen committed:
Refactor JToolbarTests

dneukirchen committed:
Refactor JVersionTest

dneukirchen committed:
Run all library tests in the library suite

dneukirchen committed:
Move JAccessTests to new unit suite

dneukirchen committed:
Refactor JAccessRulesTest

dneukirchen committed:
Add JAccessRuleTest

dneukirchen committed:
Add to string test

dneukirchen committed:
Remove JGrid (Class does not exist anymore)

dneukirchen committed:
Remove JDocumentRendererTest

JDocumentRenderer is abstract class with only a public getter

mbabker commented on composer.json:
```diff
@@ -45,6 +45,11 @@
     },
+    "autoload-dev": {
+        "psr-4": {
+            "Tests\\": "tests"
```
Why a different root namespace?  The general convention across Joomla's code (and a bulk of the PHP ecosystem) is to use your root namespace then put all tests in `\Tests` namespace below that, i.e. `Joomla\CMS\Tests`.

dneukirchen replied:
Fine for me... but then we will end up with Joomla\CMS\Tests\Unit\Libraries\Cms\Toolbar\Button...
I would prefer Joomla\Tests as root namespace.

dneukirchen replied:
changed it to Joomla\Tests

dneukirchen committed:
Rename tests/unit to tests/Unit

dgrammatiko commented:
@dneukirchen are you gonna bring the javascript tests as well? Do you want me to do aPR for that?
Is that also considered at this point?

dneukirchen commented:
@dgrammatiko yes, the plan is to move all tests back to repo. would be nice if you can do the js part. when you merge js tests back, it would be nice to update karma because of an security issue in one of its dev dependencies. tried to reach you on glip to talk about it. we are using karma 0.13, current version is 4. karma 2+ should fix it.

dgrammatiko commented:
@dneukirchen ok, I will try to do a PR this weekend

dneukirchen committed:
CS

dneukirchen committed:
Fix path to bootstrap file

TobsBobs commented:
What is the difference between this pull request and #23962?

dneukirchen commented:
* it includes more tests (re-enabled some of the old unit test which did not run for j4 before)
* tests are namespaces
* tests use namespaced classes
* tests are autoloaded
* crappy tests and integration tests in unit test suite are removed
* tests are refactored

dneukirchen committed:
Merge branch '4.0-dev' into feature/unit-tests

Hackwar commented:
While we are at it: Appveyor has a bunch of errors in its output. I doubt that it should be like that...

brianteeman commented on tests/Unit/Libraries/Cms/Access/RuleTest.php:
```diff
@@ -0,0 +1,171 @@
+ */
+class RuleTest extends UnitTestCase
+{
+	/**
```
looks like there is an extra leading tab here

infograf768 replied:
rather missing tabs below

brianteeman commented on tests/Unit/Libraries/Cms/Access/RulesTest.php:
```diff
@@ -0,0 +1,353 @@
+		$this->assertEquals($ruleIdentities, $editRule->getData());
+	}
+
+	/**
```
looks like there is an extra leading tab here

infograf768 replied:
same comment as for the other one

rdeutz commented:
look good to me

Hackwar commented:
> What is the difference between this pull request and #23962?

I'll jsut leave this here: https://github.com/joomla/test-unit/pull/9

dneukirchen committed:
CS

dneukirchen committed:
Re-enable, fix and refactor JDocumentTest

dneukirchen committed:
Refactor and re-enable JBrowserTest

dneukirchen committed:
Refactor and re-enable JObjectTest

dneukirchen committed:
Merge branch '4.0-dev' into feature/unit-tests

dneukirchen committed:
Use Joomla\Tests as root namespace

dneukirchen committed:
Refactor and re-enable JFeedParserAtomTest

dneukirchen committed:
Rename mock

dneukirchen committed:
Namespace JHtmlStringTest

dneukirchen committed:
Refactor and re-enable JFeedParserRssTest

dneukirchen committed:
Refactor and re-enable JFeedEntryTest

dneukirchen committed:
Use assertArrayHasKey

dneukirchen committed:
Improve FeedEntry tests

dneukirchen committed:
Refactor and re-enable JFeedFactoryTest

dneukirchen committed:
Refactor and re-enable JFeedLinkTest

dneukirchen committed:
Rename JFeedLinkTest To FeedLinkTest

dneukirchen committed:
Remove JFeedPersonTest

we dont test simple constructors

dneukirchen committed:
Add expectation

dneukirchen committed:
Refactor and re-enable JFeedParserTest

Hackwar commented on phpunit.xml.dist:
```diff
@@ -1,51 +1,12 @@
-			<directory>tests/unit/suites/administrator</directory>
-		</testsuite>
-		<testsuite name="FinderIndexer">
-			<directory>tests/unit/suites/finderIndexer</directory>
```
correct me if I'm wrong, but unit and integration tests will both be run by phpunit, so we have to distinguish the tests by testsuite. I also will provide tests again for the Finder Indexer, so we would have to have several runs of the test suite(s). Afaik we can only have this one phpunit.xml and also just the one bootstrap.php, so we should keep that in mind for the different tests.

dneukirchen replied:
both run by phpunit, but we can have different config and bootstrap files per test suite. 
and i would recommend this. we dont need to start the app for unit tests... 
tbh: as long as we dont have a working integration test setup, i dont care that much about the suite config in phpunit.xml. atm im not sure if we should move the integration tests to codeception or keep it in plain old phpunit...

its possible to have

`phpunit --configuration unit.phpunit.xml tests/Unit`
`phpunit --configuration integration.phpunit.xml tests/Integration`

Hackwar commented on tests/Unit/Libraries/Cms/Environment/BrowserTest.php:
```diff
@@ -0,0 +1,160 @@
+				true,
+			],
+			'Firefox 6 Windows' => [
+				'Mozilla/5.0 (Windows NT 5.0; WOW64; rv:6.0) Gecko/20100101 Firefox/6.0',
```
So my current FF version is 65.0.1... Could we update this to contain something newer than 2011?

dneukirchen replied:
can you update the datasets and open pr against the base branch? would be great!

dneukirchen committed:
CS

dneukirchen committed:
Refactor and re-enable JFeedTest

dneukirchen committed:
Refactor and re-enable JMailHelperTest

dneukirchen committed:
Fix bug in Mail class

dneukirchen committed:
Refactor and re-enable MailTest

dneukirchen committed:
Refactor and re-enable JMicrodataTest

dneukirchen committed:
Remove JPlatformTest

dneukirchen committed:
Remove tmp dir

dneukirchen committed:
Fix namespaces

dneukirchen committed:
Delete JLoaderTest

It heavily depends on filesystem and therefore should be tested with integration tests.

dneukirchen committed:
Merge branch '4.0-dev' into feature/unit-tests

dneukirchen committed:
Merge branch '4.0-dev' into feature/unit-tests

dneukirchen commented:
Rtc?

dgrammatiko commented:
@dneukirchen cross referencing 
https://github.com/joomla/joomla-cms/pull/24044

dneukirchen committed:
Merge branch '4.0-dev' into feature/unit-tests

dneukirchen committed:
Merge branch '4.0-dev' into feature/unit-tests

dneukirchen committed:
Merge branch '4.0-dev' into feature/unit-tests

dneukirchen committed:
Merge branch '4.0-dev' into feature/unit-tests

dneukirchen committed:
Merge branch '4.0-dev' into feature/unit-tests

dneukirchen committed:
Merge branch '4.0-dev' into feature/unit-tests

dneukirchen committed:
Merge branch '4.0-dev' into feature/unit-tests

dneukirchen committed:
Merge branch '4.0-dev' into feature/unit-tests

# Conflicts:
#	composer.json

dneukirchen commented:
fixed merge conflict.

dneukirchen commented:
ping @rdeutz @wilsonge @HLeithner can we get this merged?

dneukirchen committed:
Merge branch '4.0-dev' into feature/unit-tests

rdeutz committed:
Merge branch '4.0-dev' into feature/unit-tests

rdeutz reviewed (approved):
Looks good to me

dgrammatiko commented:
@rdeutz also #24044

rdeutz commented:
@dgrammatiko I am on it

infograf768 commented:
Hmm.
`tests/Unit/Libraries/Cms/Mircodata/MicrodataTest.php`

`Mircodata` ??

Also it looks like these new files have not been added as indexed to the branch.
